lib/charms/temporal_k8s/v0/temporal_host_info.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+
+from ops import (
+    ConfigChangedEvent,
+    Handle,
+    LeaderElectedEvent,
+    RelationChangedEvent,
+    RelationJoinedEvent,
+    framework,
+)
+from ops.charm import CharmBase
+from ops.framework import EventBase, EventSource, ObjectEvents
+from ops.model import ActiveStatus, Relation, WaitingStatus
+
+# The unique Charmhub library identifier, never change it
+LIBID = "024db27b47e546628c9ed7f26ddad6c8"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+RELATION_NAME = "temporal-host-info"
+
+logger = logging.getLogger(__name__)
+
+
+class TemporalHostInfoProvider(framework.Object):
+    """A class for managing the temporal-host-info interface provider."""
+
+    def __init__(self, charm: CharmBase, port: int):
+        """Create a new instance of the TemporalHostInfoProvider class.
+
+        :param: charm: The charm that is using this interface.
+        :type charm: CharmBase
+        :param: port: The port number to provide to requirers. This is typically
+            the 'frontend' service port.
+        :type port: int
+        """
+        super().__init__(charm, "temporal_host_info_provider")
+        self.charm = charm
+        self.port = port
+        charm.framework.observe(charm.on[RELATION_NAME].relation_joined, self._on_host_info_relation_changed)
+        charm.framework.observe(charm.on[RELATION_NAME].relation_changed, self._on_host_info_relation_changed)
+        charm.framework.observe(charm.on.leader_elected, self._on_config_changed)
+        charm.framework.observe(charm.on.config_changed, self._on_config_changed)
+
+    def _on_host_info_relation_changed(self, event: RelationChangedEvent | RelationJoinedEvent):
+        """Update relation data.
+
+        :param: event: The relation event that triggered this handler.
+        :type event: RelationChangedEvent | RelationJoinedEvent
+        """
+        logger.info("Handling temporal-host-info relation event")
+        if self.charm.unit.is_leader() and "frontend" in str(self.charm.config["services"]):
+            host = str(self.charm.config["external-hostname"])
+            if binding := self.charm.model.get_binding(event.relation):
+                host = host or str(binding.network.bind_address)
+            event.relation.data[self.charm.app]["host"] = host
+            event.relation.data[self.charm.app]["port"] = str(self.port)
+
+    def _on_config_changed(self, event: ConfigChangedEvent | LeaderElectedEvent):
+        """Update relation data on config change."""
+        logger.info("Config changed, updating temporal-host-info relation data")
+        if self.charm.unit.is_leader() and "frontend" in str(self.charm.config["services"]):
```

Comment:
nit: to avoid such nested if statements, i usually try to early-out if possible:

```
if not self.charm.unit.is_leader() or "frontend" not in str(self.charm.config["services"]):
    return

host = str(...)
for relation in self.charm.model.get_relations("temporal-host-info"):
    ...
```

also note: i prefer using ops `model.get_relations()` instead of doing lookup directly (`model.relations.get()`). this is because it allows ops to make changes in relation representations without needing changes in charm libs or charm code